### ISSUE TITLE
Adapter zur Steuerung eine Nibe Wärmepumpe

### ISSUE THREAD

**Issue by mguenther1** [ioBroker/AdapterRequests#422]
Nibe Wärmepumpen können über das Nibe Modbus Modul 40 ausgelesen und gesteuert werden. Alternativ funktioniert es über einen Raspberry Pi. Hier ist eine Lösung für Node-Red zu finden:
http://hausbau-becker.blogspot.com/2020/06/it-fachartikel-fernsteuerung-nibe.html

Es gibt bereits einen Nibe Wärmepumpe Adapter in ioBroker, der aber nur per Nibe Uplink die Daten ausliest. Eine Steuerung ist aber nicht möglich.

Ein Adapter, der die Steuerung übernimmt, wäre super.

**Comment by mguenther1** [comment]
kurze Ergänzung. Nach weiterem Suchen im Netz, handelt es sich wohl um einen NibePi. Diesen in ioBroker zu integrieren, wäre nen Hit. Hier der Link:
https://www.energiesparhaus.at/forum-diy-alternative-zu-nibe-modbus-modul/52722
ab Seite 3 oder 4 ist der Entwickler aus Schweden mit am Diskutieren.

**Comment by Apollon77** [comment]
Wenn es einen Adaptergibt der liesst so ist das hier doch besser als Issue bei diesem Adapter aufgehoben das der Adapter erweitert wird, oder ?!

**Comment by mguenther1** [comment]
das weiß ich eben nicht. Der bestehende Adapter basiert auf einer völlig anderen Annahme. Da wird quasi die Online-Lösung über NIBE angezapft. Über einen NibePi läuft es quasi über "modbus" direkt ohne die Hersteller-Anbindung. 
Und wie ich oben schon geschrieben habe, der aktuelle Adapter liest nur, er steuert nicht.

**Comment by m-s-b** [comment]
Mein schwedisch ist zwar denkbar schlecht, aber hier soll ein günstiger Pi ja das teure Modbus Modul von Nibe emulieren. Ist natürlich wesentlich eleganter als die (ggf. noch gebührenpflichtige) NibeUpLink Variante. Der bestehende ioBroker Adapter nutzt letzteres und das "nur" lesend. Grundsätzlich hat der Autor auch nicht ausgeschlossen die Schreibfunktion einzubauen, allerdings wenn dann eher langfristig mit niedriger Priorität. 
Meiner Meinung nach sind das halt zwei unterschiedliche Varianten die beide Ihre Daseinsberechtigung haben. Mein Favorit wäre den NibePi direkt an der Wärmepumpe und nen passenden Adapter im ioBroker, von daher +1 von meiner Seite.

**Comment by steefan85** [comment]
Ich nutze den angesprochenen NibePi; dieser hat eine Anbindung über MQTT und diese Datenpunkte lassen sich ohne Probleme in iobroker einbinden.

Das funktioniert sowohl lesend als auch schreibend und zwar für alles was auch über Modbus möglich wäre.

Nutze es für die Überwachung in iobroker und auch zur Steuerung der Lüftung hierüber.

Meiner Meinung nach ist hier kein extra Adapter notwendig.

**Comment by m-s-b** [comment]
@steefan85 
Hast du zufällig ne aktuelle Anleitung parat? Hatte das vor 2 Jahren Mal mit nodered probiert, das hat aber nicht so funktioniert. Mittlerweile läuft einiges über MQTT und ich will das Projekt noch Mal angehen.

**Comment by mguenther1** [comment]
> 



> Ich nutze den angesprochenen NibePi; dieser hat eine Anbindung über MQTT und diese Datenpunkte lassen sich ohne Probleme in iobroker einbinden.
> 
> Das funktioniert sowohl lesend als auch schreibend und zwar für alles was auch über Modbus möglich wäre.
> 
> Nutze es für die Überwachung in iobroker und auch zur Steuerung der Lüftung hierüber.
> 
> Meiner Meinung nach ist hier kein extra Adapter notwendig.

Hi Stefaan, ich habe bei mir den NibePi auch inzwischen in meine 1245 fighter gesetzt, bekomme aber das mit Node-Red und mqtt nicht hin. bist du auch im iobroker Forum unterwegs? Wenn ja, würdest du eine Anleitung posten? Ich sehe gerade m-s-b hat parallel ebenfalls danach gefragt...

**Comment by steefan85** [comment]
@m-s-b 
@mguenther1 

Sobald ich eine ruhige Minute dafür finde kann ich gerne eine kurze Anleitung hierfür erstellen; eigentlich gar keine große Sache, wenn mal einmal raus hat wie das funktioniert.

**Comment by steefan85** [comment]
@m-s-b 
@mguenther1 

Hier eine schnell zusammengestellte Anleitung; ich hoffe, sie hilft erst einmal weiter:

https://github.com/steefan85/Nibe/blob/main/nibepi2mqtt.md

**Comment by KoMa1978** [comment]
> Ich nutze den angesprochenen NibePi; dieser hat eine Anbindung über MQTT und diese Datenpunkte lassen sich ohne 
> Das funktioniert sowohl lesend als auch schreibend und zwar für alles was auch über Modbus möglich wäre.
Das habe ich bei mir auch so im Einsatz und es funktioniert soweit. 

> Meiner Meinung nach ist hier kein extra Adapter notwendig.
Das sehe ich ein wenig anders, da auf dem Pi neben dem OS auch noch NodeRED, MQTT, das Addon was mit der Nibe spricht gewartet und konfiguriert werden muss.

Wirklich schöner finde ich die Arduino Lösung. Die kleine Box arbeitet vermutlich schneller und zuverlässiger und muss nichts groß drauf haben ... Die Daten werden per UDP angenommen und per RS485 weiter gegeben.

Das finde ich deutlich eleganter gelöst als den Umweg über NodeRED und MQTT: https://www.openhab.org/addons/bindings/nibeheatpump/#arduino

**Comment by mcm1957** [comment]
Ich fasse mal zusammen:
-) Es gibt ein MODBUS Interface das mit dem modbus adapter funktionieren sollte
-) Es gibt eine Lösung mit externer Hardware die (an der Wärmepumpe) die modbus Schnittstelle verwendet.

Der Bedarf für einen dedizierten Adapter ist daher im Moment nicht erkennbar. 

Falls doch ein dediziereter Adapter benötigt wird, bitte hier antworten und eine Info warum das Modbus Interface nicht reicht hinterlassen. Issues mit dem Modbus Adapter bitte im Adapterrepository ablegen.

**Comment by KoMa1978** [comment]
Es gibt einen Modbus-Adapter von Nibe, das stimmt, nur kostet knapp 1000EUR.
Der stellt dann "richtigen" Modbus zur Verfügung über seriell, sprich benötigt noch einen "Master" (USB-Dongle, Raspi, ...).

Alles in allem keine wirklich wirtschaftliche Lösung.

Ja, die Lösung mit dem NibePI (https://github.com/anerdins/nibepi) und per MQTT funktioniert, ist allerdings (meiner Meinung nach) nicht so elegant. Die OpenHAB Anbindung mit dem NibeGW auf dem Server und einem Arduino ist da (meiner Meinung nach) deutlich eleganter und stabiler.

**Comment by steefan85** [comment]
> Es gibt einen Modbus-Adapter von Nibe, das stimmt, nur kostet knapp 1000EUR.
> 
> Der stellt dann "richtigen" Modbus zur Verfügung über seriell, sprich benötigt noch einen "Master" (USB-Dongle, Raspi, ...).
> 
> 
> 
> Alles in allem keine wirklich wirtschaftliche Lösung.
> 
> 
> 
> Ja, die Lösung mit dem NibePI (https://github.com/anerdins/nibepi) und per MQTT funktioniert, ist allerdings (meiner Meinung nach) nicht so elegant. Die OpenHAB Anbindung mit dem NibeGW auf dem Server und einem Arduino ist da (meiner Meinung nach) deutlich eleganter und stabiler.

Auch NibePi basiert auf dem angesprochenen NibeGW. Eine direkte Schnittstelle z. B. als Modbus TCP wäre natürlich eleganter bzw. schlanker.

Ziel des Projekts war allerdings nicht primär die Daten zur Verfügung zu stellen, sondern die Steuerung der Heizung über den Pi (smart price, Lüftungssteuerung usw.) zu übernehmen.

Ich nutze das nicht aber zur Stabilität kann ich nichts Negatives sagen. Absolut keine Ausfälle seit über einem Jahr.
NibeGW, NodeRED und MQTT laufen zusammen sehr zuverlässig auf einem Pi Zero 2W.

**Comment by KoMa1978** [comment]
Ja, und auch mit dem NibeGW kann man die Heizung steuern. Und auch bei mir läuft der Raspi Zero wunderbar stabil seit 1-2 Jahren inkl. der Stromversorgung über die Heizung.

Es geht mir heir tatsächlich nur um "eleganter und schlanker".

**Comment by mcm1957** [comment]
Also so ganz versteh ih das noh immer nicht:

Der Nibe modbus Adapter kostet €€€€. Ok, zu teuer,
Der NibePI  wird lt. Beschreibung an die RS485 Schnittstelle angeschlossen und an der Nibe modbus eingeschaltet.
Also braucht man nun dazu den €€€€ Modbus Adpter auch? Oder spricht due Wärmepumpe auf der RS485 ein privates Protokoll?

Icch schalte mal den issue wieder auf requested. Ev. findet sich ja wer, der das technisch im Detail klärt und/oder eine konkrete Anbinsung in iob umsetzt.

**Comment by DutchmanNL** [comment]
Ich hab mal für euch recherchiert…

Die interne Communication verläuft über eigenes geschlossenes Protocol. Mit einem EME20 (279-290€) hat man die Werte auf Modbus

diese können dan mit den Modbus Adapter ausgelesen werden oder auch Node-red (https://github.com/anerdins/node-red-contrib-nibepi)

die Node red Lib könnte auch als Basis für einen Adapter dienen

**Comment by steefan85** [comment]
> Also so ganz versteh ih das noh immer nicht:
> 
> Der Nibe modbus Adapter kostet €€€€. Ok, zu teuer, Der NibePI wird lt. Beschreibung an die RS485 Schnittstelle angeschlossen und an der Nibe modbus eingeschaltet. Also braucht man nun dazu den €€€€ Modbus Adpter auch? Oder spricht due Wärmepumpe auf der RS485 ein privates Protokoll?
> 
> Icch schalte mal den issue wieder auf requested. Ev. findet sich ja wer, der das technisch im Detail klärt und/oder eine konkrete Anbinsung in iob umsetzt.

Genau, die Nibe spricht von Haus aus kein Modbus, hat aber eine RS485 Schnittstelle.
Es gibt vom Hersteller einen Adapter, so dass diese MODBUS spricht. Dieser ist €€€€.

Für openhab gibt es eine Bibliothek NibeGW (Nibe Gateway), die mittels Arduino- oder Pi-Hardware die proprietäre Schnittstelle übersetzt und über UDP lesend/schreibend zur Verfügung stellt.

Darauf aufbauend, gibt es den NibePi, der Nibe GW nutzt und hier wiederum eine Steuerung beinhaltet sowie die Schnittstelle auch über NodeRED/MQTT zur Verfügung stellt.

In ioBroker bleibt quasi nur der bestehende Adapter der Umweg über NibePi und MQTT (ohne zusätzlichen Adapter).
(Adapter gibt es, aber Zugriff nur über das Online-Portal -> keine zufriedenstellende Lösung.)

Eleganter wäre aber die UDP-Schnitstelle des NibeGW auch als Adapter in dem ioBroker zu nutzen.

**Comment by steefan85** [comment]
> 

Das EME 20 ist eigentlich für Wechselrichter gedacht; Sunspec-Modbus Protokoll.
Das Modbus 40 Modul stellt Modbus zur Verfügung, kostet aber 700 - 1000€.

**Comment by mguenther1** [comment]
> Eleganter wäre aber die UDP-Schnitstelle des NibeGW auch als Adapter in dem ioBroker zu nutzen.

so ist es. Mit deiner Anleitung von oben habe ich es inzwischen aber hinbekommen, meine Wärmepumpe F1245 sowohl auszulesen, als auch zu steuern. Dazu hängt bei mir an der Schnittstelle (in der Wärmepumpe) ein Pi Zero2 (Spannung von der Wärmepumpe), auf dem NibePi läuft. Ein MQTT Server läuft bei mir unter ioBroker und über den MQTT Client auf der Zero werden die Daten mittels Node-Red an den ioBroker ausgetauscht. 
Hier findet man übrigens für einige Modelle, mit der der NibePi zu funktioneren scheint, auch die entsprechenden Register:
[https://github.com/anerdins/NibePi-1/tree/master/nibepi/models](https://github.com/anerdins/NibePi-1/tree/master/nibepi/models)

**Comment by steefan85** [comment]
Man kann über das Frontend des NibePi auch alle konfigurierten Register auf einmal über MQTT ausgeben lassen; dann hat man nicht den komplizierten Weg alles einzeln selbst in NodeRED konfigurieren zu müssen.
Bei mir funktioniert es aber nicht richtig und einige Register fehlen einfach. Konnte den Fehler leider nicht finden.
Am 7. Feb. 2023, 21:21 +0100 schrieb mguenther1 ***@***.***>:

**Comment by DutchmanNL** [comment]
> Eleganter wäre aber die UDP-Schnitstelle des NibeGW auch als Adapter in dem ioBroker zu nutzen.

Die broadcast MessagePad kan man auslesen muss halt die binären Daten übersetzen

**Comment by mguenther1** [comment]
> Man kann über das Frontend des NibePi auch alle konfigurierten Register auf einmal über MQTT ausgeben lassen; dann hat man nicht den komplizierten Weg alles einzeln selbst in NodeRED konfigurieren zu müssen. Bei mir funktioniert es aber nicht richtig und einige Register fehlen einfach. Konnte den Fehler leider nicht finden. Am 7. Feb. 2023, 21:21 +0100 schrieb mguenther1 ***@***.***>:
> […](#)
> > Eleganter wäre aber die UDP-Schnitstelle des NibeGW auch als Adapter in dem ioBroker zu nutzen. so ist es. Mit deiner Anleitung von oben habe ich es inzwischen aber hinbekommen, meine Wärmepumpe F1245 sowohl auszulesen, als auch zu steuern. Dazu hängt bei mir an der Schnittstelle (in der Wärmepumpe) ein Pi Zero2 (Spannung von der Wärmepumpe), auf dem NibePi läuft. Ein MQTT Server läuft bei mir unter ioBroker und über den MQTT Client auf der Zero werden die Daten mittels Node-Red an den ioBroker ausgetauscht. Hier findet man übrigens für einige Modelle, mit der der NibePi zu funktioneren scheint, auch die entsprechenden Register: https://github.com/anerdins/NibePi-1/tree/master/nibepi/models — Reply to this email directly, view it on GitHub, or unsubscribe. You are receiving this because you were mentioned.Message ID: ***@***.***>

das hatte bei mir zu einem iobroker Absturz geführt, weil ja doch in kurzer Zeit eine Menge an Daten hin und hergesendet wurden. Von daher habe ich bei mir die für mich interessanten Register per Hand eingegeben - hat zwar gedauert, aber läuft nun wie es soll

**Comment by KoMa1978** [comment]
Genau wie @mguenther1 schreibt ist es bei mir auch. PI Zero an der Heizung inkl. Stromversorgung von der Heizung.
Der RPI emuliert das 1000€ Modbus40 Modul, hat aber (meines Erachtens) den Nachteil, dass man den PI irgendwie aktuell halten sollte, wo aber leider das Image nicht mehr ganz so aktiv gepflegt wird und das man auf dem PI eben mit NodeRED konfigurieren muss, welche Messages von der Heizung zu MQTT oder von MQTT in die Heizung gebracht werden.

Schöner/Eleganter wäre ein kleiner Arduino und ein nativer Adapter, wie es in OpenHAB und ich meine auch HomeAssistant gemacht wird, der die Anfragen an bzw. Werte von dem Ardunio nimmt/weitergibt.

**Comment by steefan85** [comment]
> Die Adapter sollte dafür die selben sein und senden auf mqtt Basis wen man das kennt kan man algemein den Adapter machen

Ja, beides basiert auf NibeGW dann wäre es ideal, wenn es egal wäre, welche Hardware als Gateway genutzt wird.

**Comment by maxik690** [comment]
Hi zusammen,
ich hab mir mal ein bisschen Zeit genommen und einen ioBroker Adapter erstellt welcher die Rest API benutzt:

https://github.com/maxik690/ioBroker.nibe-rest-api

Demnächst werd ich den Adapter ins latest/beta Repository einstellen, warte da noch auf eine Rückmeldung.

muss auch noch ein bisschen an den Übersetzungen herumdoktorn :D

Bitte um Feedback! (Gern auch direkt im Repo)
